sql/src/main/java/org/apache/druid/sql/calcite/rule/FilterDecomposeConcatRule.java

```diff
@@ -0,0 +1,290 @@
+      } else if (occurrences < entry.getCount()) {
+        return impossibleMatch(nonLiterals, rexBuilder);
+      }
+    }
+
+    // Apply the regex to the matchValue to get the expected value of each non-literal.
+    final Pattern regex = Pattern.compile(regexBuilder.toString(), Pattern.DOTALL);
+    final Matcher matcher = regex.matcher(matchValue);
+    if (matcher.matches()) {
+      final List<RexNode> conditions = new ArrayList<>(nonLiterals.size());
+      for (int i = 0; i < nonLiterals.size(); i++) {
+        final RexNode operand = nonLiterals.get(i);
+        conditions.add(
+            rexBuilder.makeCall(
+                SqlStdOperatorTable.EQUALS,
+                operand,
+                rexBuilder.makeLiteral(matcher.group(i + 1))
+            )
+        );
+      }
+
+      return RexUtil.composeConjunction(rexBuilder, conditions);
+    } else {
+      return impossibleMatch(nonLiterals, rexBuilder);
+    }
+  }
+
+  /**
+   * Generate an expression for the case where matching is impossible.
+   *
+   * This expression might be FALSE and might be UNKNOWN depending on whether any of the inputs are null. Use the
+   * construct "x IS NULL AND UNKNOWN" for each arg x to CONCAT, which is FALSE if x is not null and UNKNOWN is x
+   * is null. Then OR them all together, so the entire expression is FALSE if all args are not null, and UNKNOWN if any arg is null.
+   *
+   * @param nonLiterals non-literal arguments to CONCAT
+   */
+  private static RexNode impossibleMatch(final List<RexNode> nonLiterals, final RexBuilder rexBuilder)
+  {
+    if (NullHandling.sqlCompatible()) {
+      // This expression might be FALSE and might be UNKNOWN depending on whether any of the inputs are null. Use the
+      // construct "x IS NULL AND UNKNOWN" for each arg x to CONCAT, which is FALSE if x is not null and UNKNOWN if
+      // x is null. Then OR them all together, so the entire expression is FALSE if all args are not null, and
+      // UNKNOWN if any arg is null.
+      final RexLiteral unknown =
+          rexBuilder.makeNullLiteral(rexBuilder.getTypeFactory().createSqlType(SqlTypeName.BOOLEAN));
+      return RexUtil.composeDisjunction(
+          rexBuilder,
+          Iterables.transform(
+              nonLiterals,
+              operand -> rexBuilder.makeCall(
+                  SqlStdOperatorTable.AND,
+                  rexBuilder.makeCall(SqlStdOperatorTable.IS_NULL, operand),
+                  unknown
+              )
+          )
+      );
+    } else {
+      return rexBuilder.makeLiteral(false);
+    }
+  }
+
+  /**
+   * Given a literal (which may be wrapped in a cast), remove the cast call (if any) and read it as a string.
+   * Returns null if the rex can't be read as a string.
+   */
+  @Nullable
+  private static String getAsString(final RexNode rexNode)
+  {
+    // Get matchValue from the matchLiteral (remove cast call if any, then read as string).
+    final RexNode matchLiteral = RexUtil.removeCast(rexNode);
```

Comment:
I'm not sure if this right - considering: `'asd'::integer` - if the cast is to an numeric type; does the string value is still 'asd' ?

I think the in this case the cast may only be removed if its redundant - or its to string

Reply:
The cast would always be to string, since the concat functions only accept string args. If the cast was to something else, it would have failed at the validation stage, before this rule runs. I added a comment about this.